### PR TITLE
Create handle outcome lambda

### PR DESCRIPTION
**Pull request by JoelClemence**
* Write classes to determine what the outcome was
* Build outcome object

### CODE REVIEW

File: lambdas/handle-heatmap-outcome-lambda/src/main/resources/application.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+failure:
+  reason:
+    validationFailed: Validation failed
```
Comment (galexite) [inline lambdas/handle-heatmap-outcome-lambda/src/main/resources/application.yml]: Should this be something that points directly to the input being at fault in some way, i.e., 'Invalid input'?

Comment (JoelClemence) [reply lambdas/handle-heatmap-outcome-lambda/src/main/resources/application.yml]: Fixed

File: lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionOutputFactory.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  public StepFunctionOutput createStepFunctionOutput(
+      final HeatmapRequestOutcome heatmapRequestOutcome) {
+    return Option.of(heatmapRequestOutcome)
+        .filter(this::isSuccess)
+        .toEither(heatmapRequestOutcome)
+        .fold(this::createFailedResult, this::createSucceededResult);
+  }
+
+  private StepFunctionOutput createFailedResult(HeatmapRequestOutcome failed) {
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionOutputFactory.java]: final param

File: lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionOutputFactory.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    return failedStepFunctionOutputFactory.createStepFunctionOutput(failed);
+  }
+
+  private StepFunctionOutput createSucceededResult(HeatmapRequestOutcome succeeded) {
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionOutputFactory.java]: final param

File: lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/FailedStepFunctionOutputFactory.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  private Object determineErrorForNonValidationFailure(
+      final HeatmapRequestOutcome heatmapRequestOutcome) {
+    return heatmapRequestOutcome
+        .getValidationFailure()
+        .orElseUse(heatmapRequestOutcome.getHeatmapGenerationFailure())
+        .orElseUse(heatmapRequestOutcome.getHeatmapAggregationFailure())
+        .orElse("Unknown error");
+  }
+
+  private boolean hasFailedValidation(HeatmapRequestOutcome heatmapRequestOutcome) {
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/FailedStepFunctionOutputFactory.java]: final param

File: lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplication.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.oath.cyclops.jackson.CyclopsModule;
+import java.time.Clock;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+
+@SpringBootApplication
+public class HandleHeatmapOutcomeLambdaApplication {
+
+  public static void main(String[] args) {
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplication.java]: final param

File: lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionS3ObjectKeyService.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import org.apache.commons.io.FilenameUtils;
+import org.springframework.stereotype.Component;
+import uk.gov.ukho.ais.lambda.heatmap.job.model.StepFunctionOutput;
+
+@Component
+public class StepFunctionS3ObjectKeyService {
+
+  private final Clock clock;
+
+  public StepFunctionS3ObjectKeyService(Clock clock) {
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/main/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/result/StepFunctionS3ObjectKeyService.java]: final param

File: lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplicationTest.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringBootTest
+@MockBean(AmazonS3.class)
+@TestPropertySource(properties = {"JOB_SUBMISSION_BUCKET_NAME=test-bucket"})
+public class HandleHeatmapOutcomeLambdaApplicationTest {
+
+  @Test
+  public void whenFunctionIsLoadedThenSpringContextLoads() {}
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplicationTest.java]: This is tested by the other tests so is redundant.

File: lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/JobConfigSubmissionRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        new JobConfigSubmissionRepository(
+            mockAmazonS3, mockObjectMapper, mockS3ObjectKeyService, jobSubmissionBucket);
+  }
+
+  @Test
+  public void whenSuccessfullyWritingOutputFileToS3ThenOutputReturned()
+      throws JsonProcessingException {
+    when(mockObjectMapper.writeValueAsString(stepFunctionOutput))
+        .thenReturn(stepFunctionOutputJson);
+    when(mockS3ObjectKeyService.createS3CompletedObjectKey(stepFunctionOutput))
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/JobConfigSubmissionRepositoryTest.java]: Repeating sequences of mocking code could be refactored into a setupMocksForBlah style of method?

File: lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/OutcomeHandlingFunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+          "s3://bucket/file",
+          Option.of(JobConfig.empty()),
+          Option.none(),
+          Option.none());
+
+  @Test
+  public void whenHandlingOutcomeThenCreatesOutput() {
+    SoftAssertions.assertSoftly(
+        softly -> {
+          when(mockStepFunctionOutputFactory.createStepFunctionOutput(heatmapRequestOutcome))
```
Comment (andyphelps) [inline lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/OutcomeHandlingFunctionTest.java]: As before, refactor common mocking steps into method?

File: lambdas/libraries/heatmap-job-model/src/main/java/uk.gov.ukho.ais.lambda.heatmap.job.model/HeatmapRequestOutcome.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  private final Option<Object> validationFailure;
+  private final Option<Object> heatmapGenerationFailure;
+  private final Option<Object> create6hr30kmHeatmap;
+  private final Option<Object> create18hr100kmHeatmap;
+  private final Option<Object> heatmapAggregationFailure;
+  private final Option<Object> heatmapAggregation;
+
+  @JsonCreator
+  public HeatmapRequestOutcome(
+      @JsonProperty("jobConfigFile") String jobConfigFile,
```
Comment (andyphelps) [inline lambdas/libraries/heatmap-job-model/src/main/java/uk.gov.ukho.ais.lambda.heatmap.job.model/HeatmapRequestOutcome.java]: final?

File: lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplicationTest.java (hunk 7 above)
Comment (JoelClemence) [reply lambdas/handle-heatmap-outcome-lambda/src/test/java/uk/gov/ukho/ais/lambda/handleheatmapoutcome/HandleHeatmapOutcomeLambdaApplicationTest.java]: Not really - this is the only test that loads the Spring Context and checks that it has all beans and configuration necessary

**Pull request merged**